### PR TITLE
feat(TablePropTypes) Allow non string elements in the table header

### PR DESCRIPTION
**Pull request by herleraja**
Closes https://github.com/carbon-design-system/carbon-addons-iot-react/issues/2077

**Summary**

- allow non-string elements in the table header

**Change List (commits, features, bugs, etc)**

- feat(TablePropTypes) allow non-string elements in the table header

**Acceptance Test (how to verify the PR)**

- able to view work with an existing table and other add-on components.

### CODE REVIEW

File: packages/react/src/components/Table/StatefulTable.test.jsx [hunk 1]
```diff
@@ -347,9 +347,9 @@ describe('stateful table with real reducer', () => {
     let viewProps;
     const onUserViewModified = jest
       .fn()
-      .mockImplementation(({ view, columns, state: { currentSearchValue } }) => {
+      .mockImplementation(({ view, state: { currentSearchValue } }) => {
         viewProps = {
-          columns,
+          columns: initialState.columns,
```
Comment (sstone2423) [inline packages/react/src/components/Table/StatefulTable.test.jsx:352]: what does this change?

File: packages/react/src/components/Table/TablePropTypes.js [hunk 2]
```diff
@@ -88,7 +88,7 @@ export const TableRowPropTypes = PropTypes.arrayOf(
 export const TableColumnsPropTypes = PropTypes.arrayOf(
   PropTypes.shape({
     id: PropTypes.string.isRequired,
-    name: PropTypes.string.isRequired,
+    name: PropTypes.node.isRequired,
```
Comment (sstone2423) [inline packages/react/src/components/Table/TablePropTypes.js:91]:
```suggestion
    name: PropTypes.oneOfType([PropTypes.node, PropTypes.string].isRequired,
```

Might be better to explicitly allow both strings and nodes still

Comment (herleraja) [reply packages/react/src/components/Table/TablePropTypes.js:91]: Done!

**Review by davidicus — COMMENTED** [review]
We should change the example in storybook to use the definition tooltip

**Comment by herleraja** [comment]
> We should change the example in storybook to use the definition tooltip

I tried with `TooltipDefinition` but seems like the focus was taken by header (sorting)

* custom header without focus

image

* custom header with focus (looks like we are trying too many things in the table header)
image


So I'm going to keep it the same for the time being.

File: packages/react/src/components/Table/StatefulTable.test.jsx (hunk 1 above)
Comment (herleraja) [reply packages/react/src/components/Table/StatefulTable.test.jsx]: We can not compare (isEqual) on two react component if there is a cyclic dependency(https://formidable.com/blog/2018/introducing-react-fast-compare/)

I tried to skip the test case failure by skipping the column value.

**Comment by davidicus** [comment]
@herleraja having an icon in the header probably does not allow for enough room and will provide poor usability. We need to figure out how to provide a definition tooltip to trigger the tooltip. 

image


If you look at the column in the image you can not even see the icon. There is also black text in the filter.

**Comment by herleraja** [comment]
> @herleraja having an icon in the header probably does not allow for enough room and will provide poor usability. We need to figure out how to provide a definition tooltip to trigger the tooltip.
> 
> image
> 
> If you look at the column in the image you can not even see the icon. There is also black text in the filter.

@davidicus I agree!

**Comment by herleraja** [comment]
@davidicus I think the storybook is one example. Maybe we can merge this PR so that others can customize their table header as per their requirement.

**Comment by herleraja** [comment]
Not needed anymore. 

https://github.com/carbon-design-system/carbon-addons-iot-react/pull/2272 covers the above usecase.

**Pull request closed without merging**